MLS.Agent/WebHostBuilderExtensions.cs

```diff
@@ -0,0 +1,49 @@
+// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Net;
+using System.Net.Sockets;
+using Microsoft.AspNetCore.Hosting;
+using MLS.Agent.CommandLine;
+
+namespace MLS.Agent
+{
+    public static class WebHostBuilderExtensions
+    {
+        public static IWebHostBuilder WithConfiguredApplicationUrl(this IWebHostBuilder builder, StartupOptions options)
+        {
+            Uri uri;
+
+            if (IsLaunchedForDevelopment())
+            {
+                uri = new Uri("http://localhost:4242");
+            }
+            else if (!string.IsNullOrEmpty(options.Port))
+            {
+                uri = new Uri($"https://localhost:{options.Port}");
+            }
+            else
+            {
+                uri = new Uri($"https://localhost:{GetFreePort()}");
+            }
+
+            return builder.UseUrls(uri.ToString());
+        }
+
+        private static bool IsLaunchedForDevelopment()
+        {
+            var processName = System.Diagnostics.Process.GetCurrentProcess().ProcessName;
+            return processName == "dotnet" || processName == "dotnet.exe";
+        }
+
+        private static int GetFreePort()
+        {
+            TcpListener l = new TcpListener(IPAddress.Loopback, 0);
```

Comment:
Is the second argument the port it starts searching from? Might be good to pick some value like `7000` or something so that the browser windows we open point to similar ports.

Reply:
(I'm not sure what the availability/spacing is of open ports starting at 0 on the various oses we support)

Reply:
That is the "listening socket". I tried to specify 700 in place of 0 there and it looks like if we do that it tries to open that particular port only. Specifying 0 I think implies that look for the next available free port.

Reply:
Cool, makes sense